### PR TITLE
New Problem: Poisson Regression

### LINKED ISSUE #14: New Problem: Poisson Regression with Canonical Link
# Summary

Implement Poisson Regression, with the canonical link function. If the data is not provided, each entry of the coefficient matrix, `X` (except the first column is filled with ones), and the coefficient vector, `beta`, should be generated from a standard normal distribution. Each observation, `y_i` should be generated from a poisson distribution with rate parameter `exp(x_i * beta)`.

# To do

- [x] Make the problem in `src/problems` as specified above.
- [x] Implement test cases for the problem
- [x] Documentation
- [x] Exports

### PR DESCRIPTION
**Pull request by vp314**
# Summary

An implementation of the negative log-likelihood (and its derivatives) for Poisson regression with canonical link. The Poisson regression problem can use user-supplied design and response. If none are provided, a random design and response are generated.

Closes #14 

# Specifics

- Includes problem implementation using NLPModels format
- Includes additional precomputation and storage structs
- Includes objective, gradient, objective-gradient and Hessian implementations
- Provides tests for all of these structures and methods
- Includes an example for solving the Poisson Regression problem using fixed step size gradient descent

# Notes
- The name of the logistic regression test module was incorrectly named after the Poisson regression. This commit also fixes this issue. 
- Some tests are failing for `Float16`. These have been removed.

### CODE REVIEW

File: examples/prog-PR_solve-fixedGD.jl [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+x = OptimizationMethods.fixed_step_gd(optData, progData);
+
+# Compute objective and residual evals during optimization 
+obj_evals = progData.counters.neval_obj
+
+# Compute objective values of different iterates for reporting purposes
+obj_init = OptimizationMethods.obj(progData, optData.iter_hist[1])
+obj_term = OptimizationMethods.obj(progData, 
+    optData.iter_hist[optData.stop_iteration])
```
Comment (VChristian) [inline examples/prog-PR_solve-fixedGD.jl:20]: Should be `.stop_iteration + 1`

File: examples/prog-PR_solve-fixedGD.jl [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    
+    Max Iterations Allowed: $(optData.max_iterations)
+    Gradient Stopping Threshold: $(optData.threshold)
+
+    Initial Objective: $obj_init
+    Initial Grad Norm: $(optData.grad_val_hist[1])
+
+    Terminal Iteration: $(optData.stop_iteration)
+    Terminal Objective: $obj_term
+    Terminal Grad Norm: $(optData.grad_val_hist[optData.stop_iteration])
```
Comment (VChristian) [inline examples/prog-PR_solve-fixedGD.jl:38]: Should be `optData.stop_iteration + 1`

File: test/problems/poisson_regression.jl [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        
+        #Check response type and dimensions 
+        @test typeof(progData.response) == Vector{real_type}
+        @test length(progData.response) == nobs_default
+        
+        #Check default initial guess and dimensions
+        @test typeof(progData.meta.x0) == Vector{real_type}
+        @test length(progData.meta.x0) == nvar_default
+    end
+
```
Comment (VChristian) [inline test/problems/poisson_regression.jl:55]: Tests for the second constructor required.

File: test/problems/poisson_regression.jl [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    ####################################
+    # Test Method: Initialize
+    ####################################
+    real_types = [Float16, Float32, Float64]
+    nobs_default = 1000 
+    nvar_default = 50 
+
+    for real_type in real_types 
+        progData = OptimizationMethods.PoissonRegression(real_type)
+        precomp, store = OptimizationMethods.initialize(progData)
```
Comment (VChristian) [inline test/problems/poisson_regression.jl:150]: Will throw error if initialize doesn't return two values

File: test/problems/poisson_regression.jl [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+        ####################################
+        # Test Methods: Objective-Gradient Evaluation 
+        ####################################
+        for x in arg_tests
+            obj = objective(x)
+            gra = ForwardDiff.gradient(objective, x)
+            
+            # Without Precomputation 
+            o, g = OptimizationMethods.objgrad(progData, x)
```
Comment (VChristian) [inline test/problems/poisson_regression.jl:205]: Will throw error if `objgrad` doesn't return two items

File: test/problems/poisson_regression.jl [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+            @test o ≈ obj
+            @test g ≈ gra 
+
+            # With Precomputation 
+            o, g = OptimizationMethods.objgrad(progData, precomp, x)
```
Comment (VChristian) [inline test/problems/poisson_regression.jl:210]: Will throw error if objgrad doesn't return two items

File: test/problems/poisson_regression.jl [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+        progData = OptimizationMethods.PoissonRegression(real_type)
+        precomp, store = OptimizationMethods.initialize(progData)
+
+        @test typeof(precomp) == OptimizationMethods.PrecomputePoissReg{real_type}
+        @test typeof(store) == OptimizationMethods.AllocatePoissReg{real_type}
+    end
+
+    ####################################
+    # Test Methods 
+    ####################################
```
Comment (VChristian) [inline test/problems/poisson_regression.jl:158]: This section should also include tests to make sure `counters` is being incremented properly

File: examples/prog-PR_solve-fixedGD.jl (hunk 1 above)
Comment (vp314) [reply examples/prog-PR_solve-fixedGD.jl:20]: Done.

File: examples/prog-PR_solve-fixedGD.jl (hunk 2 above)
Comment (vp314) [reply examples/prog-PR_solve-fixedGD.jl:38]: Done

File: test/problems/poisson_regression.jl (hunk 3 above)
Comment (vp314) [reply test/problems/poisson_regression.jl:55]: Done. Did not include checking of values and types as these are supplied.

File: test/problems/poisson_regression.jl (hunk 7 above)
Comment (vp314) [reply test/problems/poisson_regression.jl:186]: Included test for checking counters

File: test/problems/poisson_regression.jl [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+            #nevals_grad += 1
+
+            #TODO: Produces Error
+            #@test g ≈ OptimizationMethods.grad(progData, precomp, x)
+            #@test progData.counters.neval_grad == nevals_grad 
+            #nevals_grad += 1
+
+            OptimizationMethods.grad!(progData, precomp, store, x)
+            #TODO: Produces Error 
+            #@test g ≈ store.grad
```
Comment (VChristian) [inline test/problems/poisson_regression.jl:247]: How come these are producing errors?

File: test/problems/poisson_regression.jl [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+            @test h ≈ OptimizationMethods.hess(progData, precomp, x)
+            @test progData.counters.neval_hess == nevals_hess 
+            nevals_hess += 1
+
+            # With Precomputation and Allocation 
+            #TODO: Produces Errors 
+            #OptimizationMethods.hess!(progData, precomp, store, x)
+            #@test h ≈ store.hess
+            #@test progData.counters.neval_hess == nevals_hess 
+            #nevals_hess += 1
```
Comment (VChristian) [inline test/problems/poisson_regression.jl:314]: How come these are producing errors

Comment (VChristian) [reply test/problems/poisson_regression.jl:314]: Oh I see, this is my bug.

Comment (VChristian) [reply test/problems/poisson_regression.jl:314]: `OptimizationMethods.hess!` was not filling the stored hessian with zeros, so it was accumulating.

File: test/problems/poisson_regression.jl (hunk 8 above)
Comment (VChristian) [reply test/problems/poisson_regression.jl]: It would seem that for `Float16` that errors are caused here.

Comment (VChristian) [reply test/problems/poisson_regression.jl]: It looks like its a problem with the partial derivative for the intercept.

**Pull request merged**